### PR TITLE
Implement auto reconnect on SignalR TypeScript client

### LINKED ISSUE #8709: Support automatic reconnect in the TypeScript client
Epic #8797

Split from https://github.com/aspnet/AspNetCore/issues/5282

We will support automatic reconnect in 3.0 **without message replay**. We will also allow customization of the reconnect logic. This will be opt-in via a `HubConnectionBuilder` option. The default behavior will continue to be manual reconnection.

### PR DESCRIPTION
**Pull request by halter73**
Partially addresses #5282

Fixes #8709 (added by @anurse :trollface:)

### CODE REVIEW

**Review by bradygaster — APPROVED** [review]

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 1]
```diff
@@ -54,13 +56,19 @@ export class HttpConnection implements IConnection {
...
     private startPromise?: Promise<void>;
     private stopError?: Error;
     private accessTokenFactory?: () => string | Promise<string>;
 
     public readonly features: any = {};
     public onreceive: ((data: string | ArrayBuffer) => void) | null;
     public onclose: ((e?: Error) => void) | null;
+    public onreconnecting: ((e?: Error) => void) | null;
+
+    // REVIEW: I would like to have a connectionId param here to highlight it will change,
```
Comment (mikaelm12) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: I like this idea

File: src/SignalR/clients/ts/signalr/src/DefaultReconnectPolicy.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+import { IReconnectPolicy } from "./IReconnectPolicy";
+
+const DEFAULT_RETRY_DELAYS_IN_MILLISECONDS = [0, 2000, 10000, 30000];
+
+/** @private */
+export class DefaultReconnectPolicy implements IReconnectPolicy {
```
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/DefaultReconnectPolicy.ts]: I like this idea. That certainly seems better than adding a bunch of overloads to withAutomaticReconnect.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts (hunk 1 above)
Comment (JamesNK) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Aren't there are times when the client just won't have the connectionId? e.g. an automatic, non-negotiated (or was it no handshake?) websockets connection

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 3]
```diff
@@ -376,6 +413,83 @@ export class HttpConnection implements IConnection {
...
+
+                if (this.connectionState !== ConnectionState.Reconnecting) {
+                    return;
+                }
+            }
+
+            nextRetryDelay = reconnectPolicy.nextRetryDelayInMilliseconds(previousRetryCount++, Date.now() - startTime);
+        }
+
+        this.logger.log(LogLevel.Information, "Reconnect retry attempts have been exhausted. Connection disconnecting.");
```
Comment (JamesNK) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: IMO it would be useful info to have how many reconnect attempts before failure in the log message.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 4]
```diff
@@ -376,6 +413,83 @@ export class HttpConnection implements IConnection {
...
+            // Exit early if the onreconnecting callback called connection.stop().
+            if (this.connectionState !== ConnectionState.Reconnecting) {
+                return;
+            }
+        }
+
+        const startTime = Date.now();
+
+        while (nextRetryDelay != null) {
+            this.logger.log(LogLevel.Information, `The next reconnect attempt will start in ${nextRetryDelay} ms.`);
```
Comment (JamesNK) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: IMO it would be useful to include the retry count in the log message

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts (hunk 1 above)
Comment (mikaelm12) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Yeah, connecting with WebSockets and skipping negotiation. The client never gets a negotiation response which is where they would get their connection id from the server.  

> or was it no handshake?

Handshake is always required to connect. 😄

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 5]
```diff
@@ -54,13 +56,20 @@ export class HttpConnection implements IConnection {
     private readonly logger: ILogger;
     private readonly options: IHttpConnectionOptions;
     private transport?: ITransport;
+    private transferFormat?: TransferFormat;
     private startPromise?: Promise<void>;
     private stopError?: Error;
     private accessTokenFactory?: () => string | Promise<string>;
+    private connectionId: string | null;
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]:
```suggestion
    private connectionId?: string;
```

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts [hunk 6]
```diff
@@ -348,6 +366,26 @@ export class HubConnection {
         }
     }
 
+    /** Registers a handler that will be invoked when the connection starts reconnecting.
+     *
+     * @param {Function} callback The handler that will be invoked when the connection start reconnecting. Optionally receives a single argument containing the error that caused the connection to start reconnecting (if any).
+     */
+    public onreconnecting(callback: (error?: Error) => void) {
+        if (callback) {
+            this.reconnectingCallbacks.push(callback);
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/HubConnection.ts]: We might just want this to be a single handler that you can overwrite by calling this function again.

If we allow multiple registrations then people will want to be able to unregister.

File: src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts [hunk 7]
```diff
@@ -106,6 +107,19 @@ export class HubConnectionBuilder {
         return this;
     }
 
+     /** Configures the {@link @aspnet/signalr.HubConnection} to automatically attempt to reconnect if the connection is lost. */
+    public withAutomaticReconnect(): HubConnectionBuilder {
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts]: How common do you think people would write the following code? And accidentally overwrite their reconnect policy?
```typescript
builder.withUrl("/chat", { reconnectPolicy: new CustomReconnect() })
    .withAutomaticReconnect()
    .build();
```

File: src/SignalR/clients/ts/signalr/src/IReconnectPolicy.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+/** An abstraction that controls when the client attempts to reconnect and how many times it does so. */
+export interface IReconnectPolicy {
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/IReconnectPolicy.ts]: Need to add this to the index.ts file.

File: src/SignalR/clients/ts/signalr/src/IReconnectPolicy.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    /** Called after the transport loses the connection.
+     *
+     * @param {number} previousRetryCount The number of failed reconnect attempts so far.
+     *
+     * @param {number} elapsedMilliseconds The amount of time in milliseconds spent reconnecting so far.
+     *
+     * @returns {number | null} The amount of time to wait before the next reconnect attempt. `null` tells the client to stop retrying and close.
+     */
+    // TODO: add timeElapsedReconnecting parameter
+    nextRetryDelayInMilliseconds(previousRetryCount: number, elapsedMilliseconds: number): number | null;
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/IReconnectPolicy.ts]: We discussed making this `number | Promise<void>`, do we not want to do that anymore?

File: src/SignalR/samples/SignalRSamples/wwwroot/hubs.html [hunk 10]
```diff
@@ -164,6 +175,14 @@ <h4>Group Actions</h4>
             updateButtonState(false);
         });
 
+        connection.onreconnecting(function (e) {
+            addLine('message-list', 'Connection reconnecting: ' + e, 'red');
```
Comment (BrennanConroy) [inline src/SignalR/samples/SignalRSamples/wwwroot/hubs.html]: nit: yellow?

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 11]
```diff
@@ -376,6 +413,83 @@ export class HttpConnection implements IConnection {
...
+        }
+
+        if (!this.changeState(ConnectionState.Connected, ConnectionState.Reconnecting)) {
+            return;
+        }
+
+        if (error) {
+            this.logger.log(LogLevel.Information, `Connection reconnecting because of error '${error}'.`);
+        } else {
+            this.logger.log(LogLevel.Information, "Connection reconnecting.");
```
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Why do you like that better than "Connection reconnecting."?

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 12]
```diff
@@ -376,6 +413,83 @@ export class HttpConnection implements IConnection {
...
+            // This exception is returned to the user as a rejected Promise from the start method
+            // or was observed by the last reconnect loop.
+        }
+
+        const reconnectPolicy = this.options.reconnectPolicy as IReconnectPolicy;
+        let previousRetryCount = 0;
+        let nextRetryDelay = reconnectPolicy.nextRetryDelayInMilliseconds(previousRetryCount++, 0);
+
+        if (nextRetryDelay === null) {
+            this.logger.log(LogLevel.Information, "Connection not reconnecting because of the IReconnectPolicy.");
```
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: There's a weird situation where it's possible to have reconnectPolicy.nextRetryDelayInMilliseconds() return null the first time it's called. That's what this is handling. I don't want to log "Reconnect attempts have stopped" when the HttpConnection never entered the reconnecting state.

That being said, it is possible that the HubConnection never left the reconnecting state from the previous reconnect attempt, and this will be logged. I'll try changing the log based on the previous retry count.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts (hunk 5 above)
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: I had it that way, but then the connectionID is really `string | undefined`. Do we prefer that over `string | null`?

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts (hunk 6 above)
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HubConnection.ts]: I'm following the pattern we used for onclose(). I think if we want to change that pattern, we can change it for all onclose/onreconnecting/onreconnected all at once.

File: src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts (hunk 7 above)
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts]: Not very common, but I can see how it could happen. Do you think we should have withAutomaticReconnect no-op if a reconnectPolicy is already set?

File: src/SignalR/clients/ts/signalr/src/IReconnectPolicy.ts (hunk 9 above)
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/IReconnectPolicy.ts]: https://github.com/aspnet/AspNetCore/issues/9017

File: src/SignalR/clients/ts/signalr/src/DefaultReconnectPolicy.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/DefaultReconnectPolicy.ts]: #9016

**Comment by halter73** [comment]
@JameNK @mikaelm12 @BrennanConroy @anurse This is ready for review.

I only plan to add two more tests similar to AutomaticReconnect.test.ts's "does not reset after hub handshake failure" test that tests the integration between HubConnection and HttpConnection.

1. A test to verify the onreconnected connectionId parameter appropriately provides the new connection id.
1. A test to verify that the client can reconnect again after it successfully and fully reconnects once.

I'm also considering adding a test to verify the reconnect delay timeout gets cleared when the connection stops, but making the delay long enough to be reliably but quick enough not to slow down our tests might be tricky.

File: src/SignalR/clients/ts/signalr/tests/AutomaticReconnect.test.ts [hunk 13]
```diff
@@ -0,0 +1,449 @@
...
+
+            const startPromise = hubConnection.start();
+
+            await TestWebSocket.webSocketSet;
+            await TestWebSocket.webSocket.closeSet;
+
+            TestWebSocket.webSocket.readyState = TestWebSocket.OPEN;
+            TestWebSocket.webSocket.onopen(new TestEvent());
+
+            await new Promise((resolve) => setTimeout(resolve, 100));
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/tests/AutomaticReconnect.test.ts]: This will change to wait for onmessage to be set similar to how we wait for onclose to be set.

**Comment by halter73** [comment]
@BrennanConroy @mikaelm12 ping

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 14]
```diff
@@ -123,22 +151,55 @@ export class HttpConnection implements IConnection {
...
+            try {
+                await this.transport.stop();
+            } catch (e) {
+                this.logger.log(LogLevel.Error, `HttpConnection.transport.stop() threw error '${e}'.`);
+                this.stopConnection();
+            }
+
             this.transport = undefined;
+        } else {
+            this.logger.log(LogLevel.Debug, "HttpConnection.transport became unset because start() failed.");
```
Comment (mikaelm12) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: nit: Sounds odd. Maybe just say that it's undefined

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 15]
```diff
@@ -157,7 +218,7 @@ export class HttpConnection implements IConnection {
                     // No fallback or negotiate in this case.
                     await this.transport!.connect(url, transferFormat);
                 } else {
-                    return Promise.reject(new Error("Negotiation can only be skipped when using the WebSocket transport directly."));
```
Comment (mikaelm12) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Why did you change these? Shouldn't these continue to just reject the promise instead of throwing?

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 16]
```diff
@@ -207,12 +270,20 @@ export class HttpConnection implements IConnection {
...
+                this.connectionState = ConnectionState.Connected;
+            }
+
+            // stop() is waiting on us via this.startInternalPromise so keep this.transport around so it can clean up.
+            // This is the only case startInternal can exit in neither the connected nor disconnected state because stopConnection()
+            // will transition to the disconnected state. start() will wait for the transition using the stopPromise.
         } catch (e) {
             this.logger.log(LogLevel.Error, "Failed to start the connection: " + e);
             this.connectionState = ConnectionState.Disconnected;
+
```
Comment (mikaelm12) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: nit: Remove new line

File: src/SignalR/clients/ts/signalr/src/IConnection.ts [hunk 17]
```diff
@@ -6,6 +6,7 @@ import { TransferFormat } from "./ITransport";
 /** @private */
 export interface IConnection {
     readonly features: any;
+    readonly connectionId?: string;
```
Comment (mikaelm12) [inline src/SignalR/clients/ts/signalr/src/IConnection.ts]: Should this should be `string | null` since we are going to expose this?

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts [hunk 18]
```diff
@@ -101,48 +118,110 @@ export class HubConnection {
...
 
+        try {
+            await this.startInternal();
+
+            this.connectionState = HubConnectionState.Connected;
+            this.logger.log(LogLevel.Debug, "HubConnection connected successfully.");
+        } catch (e) {
+            this.connectionState = HubConnectionState.Disconnected;
+            this.logger.log(LogLevel.Debug, `HubConnection failed to start successfully because of error '${e}'.`);
+            return Promise.reject(e);
```
Comment (JamesNK) [inline src/SignalR/clients/ts/signalr/src/HubConnection.ts]: Because this is in an async method, should this be a throw instead of `Promise.reject(e);`

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts [hunk 19]
```diff
@@ -101,48 +118,110 @@ export class HubConnection {
...
+        } catch (e) {
+            this.logger.log(LogLevel.Debug, `Hub handshake failed with error '${e}' during start(). Stopping HubConnection.`);
+
+            this.cleanupTimeout();
+            this.cleanupPingTimer();
+
+            // HttpConnection.stop() should not complete until after the onclose callback is invoked.
+            // This will transition the HubConnection to the disconnected state before HttpConnection.stop() completes.
+            await this.connection.stop(e);
+            return Promise.reject(e);
```
Comment (JamesNK) [inline src/SignalR/clients/ts/signalr/src/HubConnection.ts]: Because this is in an async method, should this be a throw instead of `Promise.reject(e);`

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 20]
```diff
@@ -349,19 +420,33 @@ export class HttpConnection implements IConnection {
...
         // If we have a stopError, it takes precedence over the error from the transport
         error = this.stopError || error;
+        this.stopError = undefined;
+
+        if (this.connectionState === ConnectionState.Disconnected) {
+            this.logger.log(LogLevel.Debug, `Call to HttpConnection.stopConnection(${error}) was ignored because the connection is already in the disconnected state.`);
+            return;
+        }
+
+        if (this.connectionState === ConnectionState.Connecting) {
```
Comment (mikaelm12) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Is this the correct behavior? If `stopConnection `gets called while we're in the connecting state should that override connecting?

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts (hunk 18 above)
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HubConnection.ts]: I was wondering this myself. Our TypeScript client switches back-and-forth between throwing and returning rejected promises from async methods.

The only difference I can ascertain between the two is that throwing will trigger the catch of a surrounding try/catch *in the same method*. Does anyone else know of any other differences?

I switch a few returned rejected promises to throws in HttpConnection.startInternal where I need the catch block to run. Everywhere else I try to return a rejected promise since that seems to be the more common patter in our client.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts (hunk 15 above)
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: https://github.com/aspnet/AspNetCore/pull/8566#discussion_r272728004

File: src/SignalR/clients/ts/signalr/src/IConnection.ts (hunk 17 above)
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/IConnection.ts]: On HubConnection it should be.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts (hunk 20 above)
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: If `stopConnection` was called, we'd be in the Disconnecting state.

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts (hunk 18 above)
Comment (JamesNK) [reply src/SignalR/clients/ts/signalr/src/HubConnection.ts]: Something to figure out, but not in this PR 😬

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts (hunk 20 above)
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: This is what the post-startInternal logic in start() is taking care of. By the time start exits, the state will be Disconnected.

**Review by BrennanConroy — COMMENTED** [review]
I'll review more tonight/tomorrow

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 21]
```diff
@@ -349,19 +419,33 @@ export class HttpConnection implements IConnection {
...
+            this.logger.log(LogLevel.Warning, `Call to HttpConnection.stopConnection(${error}) was ignored because the connection hasn't yet left the in the connecting state.`);
```
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: I don't expect this will ever be logged since we just finished awaiting startInternal() here. startInternal() either:

1. Completes after successfully transitioning from the Connecting to Connected state.
2. Fails after moving back into the Disconnected state.
3. Sees that stop/stopInternal() transitioned the connection to the Disconnecting state and exits while still in the Disconnecting state. This is with the realization that startInternal() needs to complete for stopInternal to continue and the expectation that start() will await the stopInternal() promise and only throw after the connection has fully transition into the Disconnected state.

This check and log might be overly defensive, but the logic is complicated enough I'd like to be able to know this would be logged if and when we investigate any connection state issues.

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts [hunk 22]
```diff
@@ -61,11 +75,11 @@ export class HubConnection {
     // create method that can be used by HubConnectionBuilder. An "internal" constructor would just
     // be stripped away and the '.d.ts' file would have no constructor, which is interpreted as a
     // public parameter-less constructor.
-    public static create(connection: IConnection, logger: ILogger, protocol: IHubProtocol): HubConnection {
-        return new HubConnection(connection, logger, protocol);
+    public static create(connection: IConnection, logger: ILogger, protocol: IHubProtocol, reconnectPolicy?: IReconnectPolicy): HubConnection {
```
[the remark this answers is not in the document]
Comment (mikaelm12) [reply src/SignalR/clients/ts/signalr/src/HubConnection.ts]: Should we just add this as an additional overload instead of the breaking change

File: src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts [hunk 23]
```diff
@@ -106,6 +114,39 @@ export class HubConnectionBuilder {
...
 
+    /** Configures the {@link @aspnet/signalr.HubConnection} to automatically attempt to reconnect if the connection is lost.
+     * By default, the client will wait 0, 2, 10 and 30 seconds respectively before trying up to 4 reconnect attempts.
+     */
+    public withAutomaticReconnect(): HubConnectionBuilder;
+
+    /** Configures the {@link @aspnet/signalr.HubConnection} to automatically attempt to reconnect if the connection is lost.
+     *
+     * @param {number[]} retryDelays An array containing the delays in milliseconds before trying each reconnect attempt.
+     * The length of the array represents how many failed reconnect attempts it takes before the client will stop attempting to reconnect.
```
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts]: The array cannot contain null according to the type definition. If some tried using an array containing null anyway, the client would stop reconnecting after it reached the first null element of the array.

The DefaultReconnectPolicy ctor copies the array and adds the null at the end for you.

https://github.com/aspnet/AspNetCore/blob/ac2006e8910eec11559ecb4269512dad4703d323/src/SignalR/clients/ts/signalr/src/DefaultReconnectPolicy.ts#L14

File: src/SignalR/clients/ts/signalr/tests/HttpConnection.test.ts [hunk 24]
```diff
@@ -95,8 +81,9 @@ describe("HttpConnection", () => {
 
                 await expect(connection.start(TransferFormat.Text))
                     .rejects
-                    .toThrow("Cannot start a connection that is not in the 'Disconnected' state.");
+                    .toThrow("Cannot start an HttpConnection that is not in the 'Disconnected' state.");
             } finally {
+                (options.transport as ITransport).onclose!();
```
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/tests/HttpConnection.test.ts]: Before stop would complete before the transport closed. Now stop waits for the transport to close, so we indicate that the transport already closed before calling connection.stop() to ensure connection.stop() completes.

As for the as cast, I cant just do `options.transport!.onclose!();` because options.transport has the type `HttpTransportType | ITransport | undefined` instead of `ITransport | undefined`.

I could store the transport in a local, but meh.

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts (hunk 22 above)
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HubConnection.ts]: It's private to TypeScript. If called from JavaScript or without type checks without the last argument reconnectPolicy is just undefined I think. Same goes for the HubConnection ctor.

That's part of reason I made reconnectPolicy optional instead of creating a NullReconnectPolicy or similar.

File: src/SignalR/clients/ts/signalr/tests/HubConnection.Reconnect.test.ts [hunk 25]
```diff
@@ -0,0 +1,414 @@
...
+import { DefaultReconnectPolicy } from "../src/DefaultReconnectPolicy";
+import { HubConnection, HubConnectionState } from "../src/HubConnection";
+import { JsonHubProtocol } from "../src/JsonHubProtocol";
+
+import { VerifyLogger } from "./Common";
+import { TestConnection } from "./TestConnection";
+import {  PromiseSource } from "./Utils";
+
+describe("auto reconnect", () => {
+    it("is not enabled by default", async () => {
```
Comment (mikaelm12) [inline src/SignalR/clients/ts/signalr/tests/HubConnection.Reconnect.test.ts]: Would we want to do iterate these over the transports and protocols like we do here https://github.com/aspnet/AspNetCore/blob/d1eeb551395fa3013d7a5862bac8cdae01fff3d6/src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts#L49

File: src/SignalR/clients/ts/signalr/tests/HubConnection.Reconnect.test.ts [hunk 26]
```diff
@@ -0,0 +1,414 @@
...
+            connection.onclose!(new Error("Connection lost"));
+
+            await closePromise;
+
+            expect(onreconnectingCalled).toBe(false);
+        },
+        "Connection disconnected with error 'Error: WebSocket closed with status code: 0 ().'.");
+    });
+
+    // it("can be opted into", async () => {
```
Comment (mikaelm12) [inline src/SignalR/clients/ts/signalr/tests/HubConnection.Reconnect.test.ts]: Some testing coverage areas - Reconnect with negotiate redirects, reconnect with skip negotiate, passing in your own `IReconnectPolicy `instead of just the `DefaultReconnectPolicy`

Comment (mikaelm12) [reply src/SignalR/clients/ts/signalr/tests/HubConnection.Reconnect.test.ts]: Also being able to successfully reconnect twice.

**Review by mikaelm12 — COMMENTED** [review]
I think this looks okay for preview 4. We can refine this in the next preview as well. Let's just wrap up the testing

File: src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts]: Fair, but they can provide their own `IReconnectPolicy` and bypass the constructor https://github.com/aspnet/AspNetCore/pull/8566/files/ac2006e8910eec11559ecb4269512dad4703d323#diff-1e166c54389222a5f7d175c9a95b415bR144

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts [hunk 27]
```diff
@@ -480,27 +578,146 @@ export class HubConnection {
...
+            if (this.connectionState !== HubConnectionState.Reconnecting) {
+                this.logger.log(LogLevel.Debug, "Connection left the reconnecting state in onreconnecting callback. Done reconnecting.");
+                return;
+            }
+        }
+
+        while (nextRetryDelay !== null) {
+            this.logger.log(LogLevel.Information, `Reconnect attempt number ${previousReconnectAttempts} will start in ${nextRetryDelay} ms.`);
+
+            await new Promise((resolve) => {
```
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HubConnection.ts]: Correct. And we've moved into the disconnected state, so that's fine..

File: src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts]: In that case, what index goes goes out of bounds? Are you worried previousRetryCount or elapsedMilliseconds?

[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts]: `previousRetryCount` would go out of bounds in their array unless they specify null in the last element. Maybe that's ok, as you put a doc comment for the method, maybe we could be more explicit and say you must return a null as the last element?

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/HubConnection.ts]: That would be a memory leak then? Granted a very slow one so maybe not important enough to fix.

**Review by BrennanConroy — COMMENTED** [review]
Looks like there might be a flaky/always failing test?

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HubConnection.ts]: No. The timeout is cleared. Everything is gc'd.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 28]
```diff
@@ -349,19 +419,33 @@ export class HttpConnection implements IConnection {
...
-    private changeState(from: ConnectionState, to: ConnectionState): boolean {
-        if (this.connectionState === from) {
-            this.connectionState = to;
-            return true;
-        }
-        return false;
-    }
-
     private stopConnection(error?: Error): void {
+        this.logger.log(LogLevel.Debug, `HttpConnection.stopConnection(${error}) called while in sate ${this.connectionState}.`);
```
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: These are debug logs. If typescript made it easy to turn regular enums to strings, I would consider it as part of this change.

But as it stands, I have to either change the enum to a "string enum" or add a look up table. Not crazy, but not necessary as part of this change.

File: src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HubConnectionBuilder.ts]: I doubt most custom implementations will even use arrays. Talking about elements in a data structure people may or may not use could confuse people.

The doc comments for IReconnectPolicy indicates "`null` tells the client to stop retrying and close." I'll make sure this is clear in the main documentation as well.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts (hunk 28 above)
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Okay, I fixed it. https://github.com/aspnet/AspNetCore/pull/8566/commits/2112031bef4847b6ff046e7d3a7015b2fd0bbe1b

[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Third times a charm
"sate" -> "state"

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 29]
```diff
@@ -14,9 +14,10 @@ import { WebSocketTransport } from "./WebSocketTransport";
 
 /** @private */
 const enum ConnectionState {
-    Connecting,
-    Connected,
-    Disconnected,
+    Connecting = "Connecting ",
```
Comment (BrennanConroy) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Nice :+1:

**Review by BrennanConroy — APPROVED** [review]
Should the onclosed callback be triggered if the connection stops while in the reconnecting state?
We should file an issue for adding some random delay.
We should file an issue for figuring out how to make reconnect work nicely with handshake failure.
Some of these logs might be better suited as trace.
